drivers/i2c/i2c_wch.c

```diff
@@ -0,0 +1,498 @@
+
+	if ((data->current.msg->flags & I2C_MSG_RW_MASK) == I2C_MSG_WRITE) {
+		regs->DATAR = ((data->current.addr << 1) & 0xFF);
+	} else {
+		regs->DATAR = ((data->current.addr << 1) & 0xFF) | 1;
+		if (data->current.msg->len == 2U) {
+			regs->CTLR1 |= I2C_CTLR1_POS;
+		}
+	}
+}
+
+static void wch_i2c_handle_addr(const struct device *dev)
+{
+	const struct i2c_wch_config *config = dev->config;
+	struct i2c_wch_data *data = dev->data;
+	I2C_TypeDef *regs = config->regs;
+
+	if ((data->current.msg->flags & I2C_MSG_RW_MASK) == I2C_MSG_READ) {
+		if (data->current.msg->len == 1U) {
+			regs->CTLR1 &= ~I2C_CTLR1_ACK;
+			if (data->current.msg->flags & I2C_MSG_STOP) {
+				regs->CTLR1 |= I2C_CTLR1_STOP;
+			}
+		} else if (data->current.msg->len == 2U) {
+			regs->CTLR1 &= ~I2C_CTLR1_ACK;
+		}
+	}
+
+	regs->STAR1;
+	regs->STAR2;
+}
+
+static void wch_i2c_handle_txe(const struct device *dev)
+{
+	const struct i2c_wch_config *config = dev->config;
+	struct i2c_wch_data *data = dev->data;
+	I2C_TypeDef *regs = config->regs;
+
+	if (data->current.idx < data->current.msg->len) {
+		regs->DATAR = data->current.msg->buf[data->current.idx++];
+		if (data->current.idx == data->current.msg->len) {
+			regs->CTLR2 &= ~I2C_CTLR2_ITBUFEN;
+		}
+	} else {
+		if (data->current.msg->flags & I2C_MSG_STOP) {
+			regs->CTLR1 |= I2C_CTLR1_STOP;
+		}
+
+		if (regs->STAR1 & I2C_STAR1_BTF) {
+			regs->DATAR;
+		}
+
+		atomic_set(&data->xfer_done, 1);
+	}
+}
+
+static void wch_i2c_handle_rxne(const struct device *dev)
+{
+	const struct i2c_wch_config *config = dev->config;
+	struct i2c_wch_data *data = dev->data;
+	I2C_TypeDef *regs = config->regs;
+
+	if (data->current.idx < data->current.msg->len) {
+		if (data->current.idx == data->current.msg->len - 1) {
+			regs->CTLR1 &= ~I2C_CTLR1_ACK;
+			regs->CTLR1 &= ~I2C_CTLR1_POS;
+
+			if (data->current.msg->flags & I2C_MSG_STOP) {
+				regs->CTLR1 |= I2C_CTLR1_STOP;
+			}
+
+			regs->CTLR2 &= ~I2C_CTLR2_ITBUFEN;
+
+			data->current.msg->buf[data->current.idx++] = regs->DATAR;
+
+			atomic_set(&data->xfer_done, 1);
+		} else {
+			data->current.msg->buf[data->current.idx++] = regs->DATAR;
+		}
+	} else {
+		if (data->current.msg->flags & I2C_MSG_STOP) {
+			regs->CTLR1 |= I2C_CTLR1_STOP;
+		}
+
+		atomic_set(&data->xfer_done, 1);
+	}
+}
+
+static void wch_i2c_handle_btf(const struct device *dev)
+{
+	struct i2c_wch_data *data = dev->data;
+
+	if ((data->current.msg->flags & I2C_MSG_RW_MASK) == I2C_MSG_WRITE) {
+		wch_i2c_handle_txe(dev);
+	} else {
+		wch_i2c_handle_rxne(dev);
+	}
+}
+
+static void i2c_wch_event_isr(const struct device *dev)
+{
+	const struct i2c_wch_config *config = dev->config;
+	struct i2c_wch_data *data = dev->data;
+	I2C_TypeDef *regs = config->regs;
+	uint16_t status = regs->STAR1;
+	bool write;
+
+	write = ((data->current.msg->flags & I2C_MSG_RW_MASK) == I2C_MSG_WRITE);
+
+	if (status & I2C_STAR1_SB) {
+		wch_i2c_handle_start_bit(dev);
+	} else if (status & I2C_STAR1_ADDR) {
```

Comment:
Are these status bits exclusive? i.e. can you use `else if`?

Reply:
If SB for example comes, we need to wait for ADDR interrupt, even if the ADDR bits are set AFAIU, so `else if` is good here.